C++ modernize: Replace MIN/MAX with type safe std::min/std::max

kvakvs opened the pull request:
- Updates C library include lines to instead include C++ wrappers
- Includes `<algorithm>` and replaces `MIN` and `MAX` macros with type safe `std::min` and `std::max`

stefanos82 commented on tests/libtransmission/copy-test.cc:
```diff
@@ -57,7 +59,7 @@ class CopyTest : public SandboxedTest
         while (buf_pos < buf_len && bytes_remaining > 0)
         {
-            uint64_t const chunk_size = MIN(buf_len - buf_pos, bytes_remaining);
+            uint64_t const chunk_size = std::min(buf_len - buf_pos, bytes_remaining);
```
You forgot to wrap `buf_len - buf_pos` in parentheses and type cast it to `uint64_t`; in other words, it should become `std::min((uint64_t)(buf_len - buf_pos), bytes_remaining);`

kvakvs replied:
Left argument both operands are `size_t` same as `uint64_t` on 64-bit. Hence i got no compiler warning on that.
Fixed now.

ckerr commented on libtransmission/announcer.cc:
```diff
@@ -6,10 +6,11 @@
-#include <stdio.h>
-#include <stdlib.h> /* qsort() */
-#include <string.h> /* strcmp(), memcpy(), strncmp() */
+#include <climits> /* INT_MAX */
+#include <cstdio>
+#include <cstdlib> /* qsort() */
+#include <cstring> /* strcmp(), memcpy(), strncmp() */
+#include <algorithm>
```
here and elsewhere, headers in an `#include` group should be sorted

kvakvs replied:
Tis done.

ckerr reviewed (approved):
LGTM. There are a few conflicts with `master` at the moment but once those are fixed this should be ready

kvakvs commented:
In all std::min/max invocations now, where the type is known, i specify that type as template parameter: `std::min<uint64_t>` like so, this pins the arguments of the template functions to be `uint64_t`. For invocations where two different types come in, instead of casting, i specify the largest integer type, this will force the compiler to up-cast the smaller value to the larger type. Please have a look again @ckerr

mikedld commented:
This last change doesn't sounds like a good idea. When some value is being truncated, neither GCC nor Clang issue a warning with `-Wall` since those don't include `-Wconversion` and for a good reason IMHO (it results in a lot of warnings nobody usually wants to spend time on). Casting arguments themselves is better, although not with `static_cast<T>(x)` which would just suppress the potential issue but with `T{x}` which would result in `-Wnarrowing` (GCC) or `-Wc++11-narrowing` (Clang), and those are in fact included in `-Wall`.

kvakvs commented:
@mikedld Nice, learned something today. Removed the template parameters and replaced would-be C-style casts with `integertype { value }` style casts. CLion instantly highlights where it would create an error.

stefanos82 commented:
> This last change doesn't sounds like a good idea. When some value is being truncated when passed in as an argument (which is what might happen with explicit template parameter specification), neither GCC nor Clang issue a warning with `-Wall` since those don't include `-Wconversion` and for a good reason IMHO (it results in a lot of warnings nobody usually wants to spend time on). Casting arguments themselves is better, although not with `static_cast<T>(x)` which would just suppress the potential issue but with `T{x}` which would result in `-Wnarrowing` (GCC) or `-Wc++11-narrowing` (Clang), and those are in fact included in `-Wall`.

I have checked the enumeration valuess where they get assigned and 1 is assigned to a type `time_t`, 2 to type `size_t` and the rest are of type `int`.

My suggestion would be to separate those different types to either their own `enum : <type> {}` or `constexpr <type> ...`.

Example:

```c++

enum : int {};
enum : time_t {};
enum : size_t {};
```

This way no more issues with explicit type casting, thus no more warnings of mismatched types during `std::min()/std::max()`.

kvakvs commented:
> My suggestion would be to separate those different types to either their own `enum : <type> {}` or `constexpr <type> ...`.
> 
> Example:
> 
> ```c++
> enum : int {};
> enum : time_t {};
> enum : size_t {};
> ```
> 
> This way no more issues with explicit type casting, thus no more warnings of mismatched types during `std::min()/std::max()`.

I am happy to convert enums to `enum class` syntax, when we visit the files with those enums. This can be added as a note to the modernization guidelines in `libtransmission/README.md`

stefanos82 commented:
> I am happy to convert enums to `enum class` syntax, when we visit the files with those enums. This can be added as a note to the modernization guidelines in `libtransmission/README.md`

Yes please!

ckerr commented:
It might be better to take this PR even with these two new lgtm warnings.

I _think_ these are actual issues and adding `FIXME` comments is valid here, since the real fix would take a refactor that's out-of-scope for a s/MIN/std::min/ PR.

Does anyone object to this being merged?

stefanos82 commented:
> It might be better to take this PR even with these two new lgtm warnings.
> 
> I _think_ these are actual issues and adding `FIXME` comments is valid here, since the real fix would take a refactor that's out-of-scope for a s/MIN/std::min/ PR.
> 
> Does anyone object to this being merged?

Fine by me; go for it.

mikedld commented on libtransmission/announcer.cc:
```diff
@@ -6,11 +6,11 @@
+#include <algorithm>
+#include <climits> /* INT_MAX */
+#include <cstdio>
+#include <cstdlib> /* qsort() */
```
Strictly speaking, including C++ counterpart of C header only provides access to declarations in `std::` namespace, not to the ones in global namespace that C header would provide. Basically, you can't call `qsort()` (`::qsort()`) after this and have to use `std::qsort()` instead. Also, these changes have nothing to do with MIN/MAX fixes...

stefanos82 replied:
Well, eventually these header files might get removed and be replaced with actual C++ header files, such as [<limits>](https://en.cppreference.com/w/cpp/header/limits), `std::qsort` with actual `std::sort`, and so forth.

kvakvs replied:
Eventually each `*.cc` file will be visited and all std library calls will have `std::` prefixes or rewritten to new C++ semantic. See my other draft pull request it does exactly that. I am not continuing on #1802 till we merge this and the modernization steps #1801, because they base one on another.

mikedld replied:
I'm not saying we shouldn't do this. What I'm trying to say is that this change is not only irrelevant to the PR topic, but is also incomplete. Something to keep in mind for your future PRs, I guess.

mikedld commented on libtransmission/tr-getopt.cc:
```diff
@@ -100,21 +101,23 @@ static void getopts_usage_line(tr_option const* opt, int longWidth, int shortWid
     if (o->longName != nullptr)
     {
-        *longWidth = MAX(*longWidth, (int)strlen(o->longName));
+        *longWidth = std::max((size_t)*longWidth, size_t{ strlen(o->longName) });
```
Note the "here _and below_" above.

kvakvs replied:
Fixed the "and below" too

mikedld reviewed (approved):
Although I would normally not approve a PR that declares one thing but then does another inside, either instead or in addition to, I guess it's better to keep it moving for the time being while the interest is there...

kvakvs commented:
I'm the kind of person that loves doing 30000 lines pull requests.
I know it is very wrong and hard to review :) Working on that negative trait of myself.
Oil this review machine better and you'll have modernized code in no time.

stefanos82 commented:
Can we take the discussion in #1807 ?

I would like to suggest a couple of things if I may; here is not the right place for long discussions.
